trulens_eval/trulens_eval/feedback/provider/cortex.py

```diff
@@ -0,0 +1,105 @@
+       Reference: https://docs.snowflake.com/en/sql-reference/functions/complete-snowflake-cortex
+    """
+
+    endpoint: CortexEndpoint
+    snowflake_session: Session
+
+    def __init__(
+        self,
+        model_engine: Optional[str] = None,
+        snowflake_connection_params: Optional[Dict] = None,
+        *args,
+        **kwargs: dict
+    ):
+        self_kwargs = dict(kwargs)
+
+        self_kwargs['model_engine'] = self.DEFAULT_MODEL_ENGINE if model_engine is None else model_engine
+        self_kwargs["endpoint"] = CortexEndpoint(
+            *args, **kwargs
+        )
+
+        connection_params = {
+            "account": os.environ["SNOWFLAKE_ACCOUNT"],
+            "user": os.environ["SNOWFLAKE_USER"],
+            "password": os.environ["SNOWFLAKE_USER_PASSWORD"],
+        } if snowflake_connection_params is None else snowflake_connection_params
+
+        # Create a Snowflake session
+        self_kwargs['snowflake_session'] = Session.builder.configs(
+            connection_params).create()
+
+        super().__init__(**self_kwargs)
+
+    def _escape_string_for_sql(self, input_string: str) -> str:
+        escaped_string = input_string.replace('\\', '\\\\')
+        escaped_string = escaped_string.replace("'", "''")
+        return escaped_string
+
+    def _exec_snowsql_complete_command(self, model: str, temperature: float, messages: Optional[Sequence[Dict]] = None):
+        # Ensure messages are formatted as a JSON array string
+        if messages is None:
+            messages = []
+        messages_json_str = json.dumps(messages)
+
+        options = {'temperature': temperature}
+        options_json_str = json.dumps(options)
+
+        completion_input_str = f"""SELECT SNOWFLAKE.CORTEX.COMPLETE(
+            '{model}',
+            parse_json('{self._escape_string_for_sql(messages_json_str)}'),
+            parse_json('{self._escape_string_for_sql(options_json_str)}')
+        )"""
+
+        # Executing Snow SQL command requires an active snow session
+        return self.snowflake_session.sql(completion_input_str).collect()
+
+    def _create_chat_completion(
+        self,
+        prompt: Optional[str] = None,
+        messages: Optional[Sequence[Dict]] = None,
+        **kwargs
+    ) -> str:
+        if 'model' not in kwargs:
+            kwargs['model'] = self.model_engine
+        if 'temperature' not in kwargs:
+            kwargs['temperature'] = 0.0
+
+        if messages is not None:
+            kwargs['messages'] = messages
+
+        elif prompt is not None:
+            kwargs['messages'] = [
+                {
+                    "role": "system",
+                    "content": prompt
+                }
+            ]
+        else:
+            raise ValueError("`prompt` or `messages` must be specified.")
+
+        res = self._exec_snowsql_complete_command(**kwargs)
+
+        completion = json.loads(res[0][0])["choices"][0]["messages"]
```

Comment:
what else is being returned?

Reply:
see example response from the notebook:
```
[Row(SNOWFLAKE.CORTEX.COMPLETE(
            'SNOWFLAKE-ARCTIC',
            [
            {'ROLE': 'USER', 'CONTENT': '
         WE HAVE PROVIDED CONTEXT INFORMATION BELOW. 
            [[''\\NTHE UNIVERSITY OF WASHINGTON, FOUNDED IN 1861 IN SEATTLE, IS A PUBL='{\n  "choices": [\n    {\n      "messages": " The University of Washington was founded in 1861."\n    }\n  ],\n  "created": 1718315813,\n  "model": "snowflake-arctic",\n  "usage": {\n    "completion_tokens": 14,\n    "prompt_tokens": 148,\n    "total_tokens": 162\n  }\n}')]
 ``` 

ref doc: https://docs.snowflake.com/en/sql-reference/functions/complete-snowflake-cortex#controlling-temperature-and-tokens

Reply:
please add this as a comment.

Reply:
Also we should not "throw-away" other params like total tokens, prompt_tokens etc. ?

Reply:
@sfc-gh-pdharmana those params are being captured by the cortex endpoint callback.

see `handle_generation` in `class CortexCallback(EndpointCallback)`